### PR TITLE
feat(profiling): use fallback 2d canvas renderer for flamechart

### PR DESCRIPTION
**Pull request by JonasBa**
The profiling UI has currently been unavailable when a WebGL canvas context could not be obtained, this change implements a fallback approach that uses a 2D canvas renderer. The downside of the 2D canvas renderer is that it is much slower than the GL version as we are bottlenecked by the limitation of making 1 draw call per rectangle we are rendering, making this a degraded experience (however still usable for most flamecharts/graphs). 

The 2D renderer is missing draw call optimizations and will currently draw everything regardless of the user fov (I will add this optimization later) and we also need to add a 2D renderer for UI frames (which currently uses GL) to have this work on all platforms.

I am debating whether to add a "your browser does not support WebGL, this is a degraded experience" type banner in the UI. Since users can't really do anything about this besides changing browsers or enabling/disabling security or privacy features (which like to break WebGL on purpose), I am tempted to just transparently fallback to the 2D renderer.

To summarize the next steps
- [ ] add a 2D UI frames renderer and fallback to it the same way as the flamegraph renderer
- [ ] add fov optimization to the 2d canvas renderer
- [ ] possibly add the "degraded experience banner"?

### CODE REVIEW

File: static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx [hunk 1]
```diff
@@ -169,10 +172,25 @@ export function AggregateFlamegraph(props: AggregateFlamegraphProps): ReactEleme
...
+        flamegraphTheme,
+        {
+          colorCoding,
+          draw_border: true,
+        },
+      ]
+    );
+
+    if (renderer === null) {
+      Sentry.captureException('Failed to initialize a flamegraph renderer');
```
Comment (Zylphrex) [inline static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx:189]: Can we potentially show an error message somewhere?

File: static/app/components/profiling/flamegraph/differentialFlamegraph.tsx [hunk 2]
```diff
@@ -182,15 +185,25 @@ export function DifferentialFlamegraph(props: DifferentialFlamegraphProps): Reac
...
+        flamegraphTheme,
+        {
+          colorCoding,
+          draw_border: true,
+        },
+      ]
     );
+
+    if (renderer === null) {
+      Sentry.captureException('Failed to initialize a flamegraph renderer');
```
Comment (Zylphrex) [inline static/app/components/profiling/flamegraph/differentialFlamegraph.tsx:202]: Showing an error would be nice here too

File: static/app/components/profiling/flamegraph/flamegraph.tsx [hunk 3]
```diff
@@ -1163,10 +1165,25 @@ function Flamegraph(): ReactElement {
...
+        flamegraphTheme,
+        {
+          colorCoding,
+          draw_border: true,
+        },
+      ]
+    );
+
+    if (renderer === null) {
+      Sentry.captureException('Failed to initialize a flamegraph renderer');
```
Comment (Zylphrex) [inline static/app/components/profiling/flamegraph/flamegraph.tsx:1182]: Showing an error would be nice here too

File: static/app/components/profiling/flamegraph/flamegraphZoomViewMinimap.tsx [hunk 4]
```diff
@@ -76,12 +79,22 @@ function FlamegraphZoomViewMinimap({
...
+      [
+        flamegraphMiniMapCanvasRef,
+        flamegraph,
+        flamegraphTheme,
+        {colorCoding, draw_border: false},
+      ]
     );
+
+    if (renderer === null) {
+      Sentry.captureException('Failed to initialize a flamegraph renderer');
```
Comment (Zylphrex) [inline static/app/components/profiling/flamegraph/flamegraphZoomViewMinimap.tsx:93]: Showing an error would be nice here too

File: static/app/utils/profiling/gl/utils.ts [hunk 5]
```diff
@@ -14,6 +17,22 @@ import {SpanChartRenderer2D} from '../renderers/spansRenderer';
 import {SpanChartNode} from '../spanChart';
 import {Rect} from '../speedscope';
 
+export function initializeFlamegraphRenderer(
+  renderers: FlamegraphRendererConstructor[],
+  constructorArgs: ConstructorParameters<FlamegraphRendererConstructor>
+): FlamegraphRenderer | null {
+  for (const renderer of renderers) {
+    const r = new renderer(...constructorArgs);
```
Comment (Zylphrex) [inline static/app/utils/profiling/gl/utils.ts:25]: Won't this throw in some cases and never try to next renderer class?

Comment (JonasBa) [reply static/app/utils/profiling/gl/utils.ts:25]: Hmm, it's possible. Let me wrap it in a try/catch

File: static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx (hunk 1 above)
Comment (JonasBa) [reply static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx:189]: I'll see if we can add an overlay to the flamegraph canvas, else I think a toast should be enough. We should never enter this case with the fallback renderers

Comment (Zylphrex) [reply static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx:189]: Im fine with a toast here 👍

**Review by Zylphrex — APPROVED** [review]
